src/js/osweb/items/loop.js

```diff
@@ -251,14 +251,14 @@ export default class Loop extends Item {
       // Create a live matrix that takes into account the repeats and the
       // shuffles.
       this.matrix = []
-      this._cycles = []
       for (let k in cycles) {
         this.matrix.push(this.orig_matrix[cycles[k]])
-        this._cycles.push(k)
       }
-      // Perform the operations
+      // Perform the operations. This may change the number of cycles, which
+      // is why this._cycles is only determined afterwards.
       this.matrix = this._operations.reduce((mtrx, [func, args]) =>
-        func(mtrx, ...this._eval_args(args)), this.matrix)
+        func(mtrx, ...this._eval_args(args)), this.matrix)      
+      this._cycles = Array(... this.matrix.keys())

```

Comment:
Shouldn't just `this.matrix.keys()` not suffice because it yields the same result as `Array(...this.matrix.keys())` or `[...this.matrix.keys()]`?

Reply:
`Array.keys()` gives an iterator that lacks functions such as `shift()` which are needed. Changing it to an `Array` like this seemed the strange-but-common solution to this. Or am I missing something? (I'm still learning the basics of JavaScript, so don't assume that my edits make sense.)

Reply:
No, I didn't know that. Then this seems perfectly fine. The canonical way of doing this nowadays is `[...this.matrix.keys()]`, but using the `Array()` constructor is not wrong either.